### PR TITLE
[commands] prefix commands for easier categorising and searching 

### PR DESCRIPTION
**Pull request by vince-fugnitto**
**Prefixed** commands in order to categorise and search commands through the `quick open command palette` more easily. 

Based on discussions from https://github.com/theia-ide/theia/issues/3287#issuecomment-433639535

**Examples:**
screen shot 2018-10-29 at 3 12 23 pm
screen shot 2018-10-29 at 3 13 32 pm

### CODE REVIEW

File: packages/core/src/common/command.ts [hunk 1]
```diff
@@ -36,6 +36,10 @@ export interface Command {
      * An icon class of this command.
      */
     iconClass?: string;
+    /**
+     * A prefix of this command.
+     */
+    prefix?: string;
```
Comment (akosyakov) [inline packages/core/src/common/command.ts]: Do we need it? How about just adding a category for commands which does not have it yet?

Also, It should be called `category` not a prefix if we introduce it.

Comment (akosyakov) [reply packages/core/src/common/command.ts]: Also prefixing does not make sense always, see discussions here: https://github.com/theia-ide/theia/issues/1119#issuecomment-372812915

File: packages/core/src/browser/quick-open/quick-command-service.ts [hunk 2]
```diff
@@ -59,6 +59,12 @@ export class QuickCommandService implements QuickOpenModel, QuickOpenHandler {
     getOptions(): QuickOpenOptions {
         return { fuzzyMatchLabel: true };
     }
+
+    private compareCommands(a: Command, b: Command): number {
```
Comment (kittaakos) [inline packages/core/src/browser/quick-open/quick-command-service.ts]: Let's move this as a comparator function under the `Command` namespace? What do you think?

At least, please make it `undefined`-safe. Yes, I know, it wasn't `undefined`-safe earlier either but it was a local code, now, you moved the logic into a method. Please fix it.

File: packages/core/src/common/command.ts (hunk 1 above)
Comment (vince-fugnitto) [reply packages/core/src/common/command.ts]: > Do we need it? How about just adding a category to command labels which don't have it yet?

I tried simply adding the category to the command labels but this had the unwanted side-effect of also affecting the menu item labels. The easiest way was for me to add a new field, and only in the display of the `quick-command-palette` would I prefix the commands with their category.

> Also, It should be called `category` not a prefix if we introduce it.

Sure, I'll make the change :)

File: packages/core/src/browser/quick-open/quick-command-service.ts (hunk 2 above)
Comment (vince-fugnitto) [reply packages/core/src/browser/quick-open/quick-command-service.ts]: > Let's move this as a comparator function under the `Command` namespace? What do you think?
> 
> At least, please make it `undefined`-safe. Yes, I know, it wasn't `undefined`-safe earlier either but it was a local code, now, you moved the logic into a method. Please fix it.

Sure @kittaakos, I'll make the necessary changes

File: packages/core/src/common/command.ts (hunk 1 above)
Comment (vince-fugnitto) [reply packages/core/src/common/command.ts]: > Also prefixing does not make sense always, see discussions here: [#1119 (comment)](https://github.com/theia-ide/theia/issues/1119#issuecomment-372812915)

I tried to base myself as much as I could on how `vscode` prefixes their commands. For instance, if we prefix all `Terminal` commands, it makes it a lot easier in the future to add more Terminal commands that a user can easily search by. However, we can go on a case-by-case basis and determine what makes sense and what doesn't. I'll make another passthrough all the commands.

**Comment by vince-fugnitto** [comment]
@akosyakov @kittaakos I've updated the PR based on your comments :)

**Comment by akosyakov** [comment]
@vince-fugnitto the important thing that it should not break the old approach for external extensions, i will verify it tomorrow

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto the important thing that it should not break the old approach for external extensions, i will verify it tomorrow

Sure no problem!

**Comment by akosyakov** [comment]
Having category seems to be totally optional and the old approach still works.

I've noticed that some commands look repetitive though:
screen shot 2018-11-01 at 09 47 18
screen shot 2018-11-01 at 09 47 08

I've not looked thoroughly through all changed commands.

**Comment by vince-fugnitto** [comment]
> Having category seems to be totally optional and the old approach still works.
> I've not looked thoroughly through all changed commands.

thanks @akosyakov, I'll make another passthrough to see if everything makes sense, I must have missed the terminal one because of how its not visible until you have an active terminal with content. 

For something that fuzzy picks up, what would be the best behaviour? 
This is `vscode`, they don't seem to have a `search in workspace` but rather `find in files`.
screen shot 2018-11-01 at 5 14 23 am

**Comment by akosyakov** [comment]
`find in files` sounds fine to me, @svenefftinge ?

**Comment by vince-fugnitto** [comment]
> `find in files` sounds fine to me, @svenefftinge ?

@akosyakov I've updated the PR to change the label to `Find in Files`, let me know if there any other changes required :)

**Pull request merged**